File: lifecycle/savedstate-core/src/main/java/androidx/lifecycle/SavedStateRegistry.java
/*
 * Copyright 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.lifecycle;

import android.os.Bundle;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * An interface for plugging components that consumes and contributes to the saved state.
 *
 * @param <S> represents a class for saving a state, typically it is {@link Bundle}
 */
public interface SavedStateRegistry<S> {
    /**
     * Consumes saved state previously supplied by {@link SavedStateProvider} registered
     * via {@link #registerSavedStateProvider(String, SavedStateProvider)}
     * with the given {@code key}.
     * <p>
     * This call clears an internal reference to returned saved state, so if you call it second time
     * in the row it will return {@code null}.
     * <p>
     * All unconsumed values will be saved during {@code onSaveInstanceState(Bundle savedState)}
     * <p>
     * This method can be called after {@code super.onCreate(savedStateBundle)} of the corresponding
     * component. Calling it before that will result in {@code IllegalArgumentException}.
     * {@link Lifecycle.Event#ON_CREATE} can be used as a signal
     * that a saved state can be safely consumed.
     *
     * @param key a key with which {@link SavedStateProvider} was previously registered.
     * @return {@code S} with the previously saved state or {@code null}
     */
    @MainThread
    @Nullable
    S consumeRestoredStateForKey(@NonNull String key);

    /**
     * Returns if a state was restored and can be safely consumed
     * with {@link #consumeRestoredStateForKey(String)}
     *
     * @return true if state was restored.
     */
    @MainThread
    boolean isRestored();

    /**
     * This interface marks a component that contributes to saved state.
     *
     * @param <S> represents a class for saving a state, typically it is {@link Bundle}
     */
    interface SavedStateProvider<S> {
        /**
         * Called to retrieve a state from a component before being killed
         * so later the state can be received from {@link #consumeRestoredStateForKey(String)}
         *
         * @return Bundle with your saved state.
         */
        @NonNull
        S saveState();
    }

    /**
     * Registers a {@link SavedStateProvider} by the given {@code key}. This
     * {@code savedStateProvider} will be called
     * during state saving phase, returned bundle will be associated with the given {@code key}
     * and can be used after the restoration via {@link #consumeRestoredStateForKey(String)}.
     * <p>
     * If there is unconsumed value with the same {@code key},
     * the value supplied by {@code savedStateProvider} will be override and
     * will be written to resulting saved state bundle.
     * <p> if a provider was already registered with the given {@code key}, an implementation should
     * throw an {@link IllegalArgumentException}
     * @param key a key with which returned saved state will be associated
     * @param savedStateProvider savedStateProvider to get saved state.
     */
    @MainThread
    void registerSavedStateProvider(@NonNull String key,
            @NonNull SavedStateProvider<S> savedStateProvider);

    /**
     * Unregisters a component previously registered by the given {@code key}
     *
     * @param key a key with which a component was previously registered.
     */
    @MainThread
    void unregisterSavedStateProvider(@NonNull String key);
}

File: savedstate/savedstate/src/commonMain/kotlin/androidx/savedstate/SavedStateRegistry.kt
/*
 * Copyright 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.savedstate

import androidx.annotation.MainThread
import androidx.savedstate.internal.SavedStateRegistryImpl

/**
 * Registry for components that consume and contribute to saved state.
 *
 * Use this registry to save and restore component state across process death or recreation.
 */
public expect class SavedStateRegistry : SavedStateProvider, SavedStateRestorer {

    /** Creates an empty [SavedStateRegistry]. */
    public constructor()

    /**
     * Creates a [SavedStateRegistry] initialized with [initialState].
     *
     * @param initialState The initial saved state to restore from.
     */
    public constructor(initialState: SavedState?)

    internal constructor(impl: SavedStateRegistryImpl)

    override fun saveState(): SavedState

    override fun restoreState(savedState: SavedState?)

    /**
     * Contributes to the saved state.
     *
     * Implementations can optionally implement [SavedStateRestorer] to receive and restore state
     * during the state restoration phase.
     */
    public fun interface SavedStateProvider : androidx.savedstate.SavedStateProvider {
        /**
         * Called to retrieve the state from a component before it is killed so the state can be
         * retrieved later from [consumeRestoredStateForKey].
         *
         * @return The [SavedState] containing the saved state.
         */
        override fun saveState(): SavedState
    }

    /**
     * Returns `true` if state has been restored and can be safely consumed with
     * [consumeRestoredStateForKey], `false` otherwise.
     */
    public val isRestored: Boolean

    /**
     * Consumes the saved state previously supplied by a [SavedStateProvider] registered with the
     * given [key].
     *
     * If the registered [SavedStateProvider] implements [SavedStateRestorer], the state is restored
     * automatically during restoration, and calls to this method with the same key return `null`.
     *
     * This call clears the internal reference to the returned saved state. Subsequent calls with
     * the same key return `null`.
     *
     * All unconsumed values are preserved during state saving.
     *
     * @param key The key with which the [SavedStateProvider] was previously registered.
     * @return The previously saved state, or `null` if none exists or it has already been consumed.
     */
    @MainThread public fun consumeRestoredStateForKey(key: String): SavedState?

    /**
     * Registers a [SavedStateProvider] with the given [key].
     *
     * This [SavedStateProvider] will be called during state saving. The returned state is
     * associated with the given [key] and can be consumed after restoration via
     * [consumeRestoredStateForKey].
     *
     * If the registered [provider] implements [SavedStateRestorer], its
     * [SavedStateRestorer.restoreState] method is automatically invoked during state restoration,
     * or immediately if state has already been restored.
     *
     * If a provider was already registered with the given [key], it is replaced with the new
     * [provider].
     *
     * @param key The key to associate with the provider.
     * @param provider The [SavedStateProvider] to register.
     */
    @MainThread public fun registerSavedStateProvider(key: String, provider: SavedStateProvider)

    /**
     * Returns the [SavedStateProvider] previously registered with [registerSavedStateProvider], or
     * `null` if no provider has been registered with the given [key].
     *
     * @param key The key used to register the [SavedStateProvider].
     */
    public fun getSavedStateProvider(key: String): SavedStateProvider?

    /**
     * Unregisters a component previously registered with the given [key].
     *
     * @param key The key with which the component was previously registered.
     */
    @MainThread public fun unregisterSavedStateProvider(key: String)
}
